### DISCUSSION TITLE
@koa/router v11+ benchmarks 10x SLOWER than v10!

### DISCUSSION THREAD

**Discussion by jeremy-daley-kr** [koajs/koa#1781] [General]
Don't just take my word for it... Run this benchmark: https://github.com/delvedor/router-benchmark

These are the time differences I'm seeing from v10 to v11:

```sh
v11

======================
 koa-router benchmark
======================
short static: 472,156 ops/sec
static with same radix: 480,787 ops/sec
dynamic route: 466,452 ops/sec
mixed static dynamic: 474,721 ops/sec
long static: 482,460 ops/sec
wildcard: 480,499 ops/sec
all together: 80,170 ops/sec

v10

======================
 koa-router benchmark
======================
short static: 4,301,032 ops/sec
static with same radix: 4,178,987 ops/sec
dynamic route: 4,033,963 ops/sec
mixed static dynamic: 3,810,142 ops/sec
long static: 3,998,437 ops/sec
wildcard: 4,125,847 ops/sec
all together: 625,066 ops/sec
```

Even v7-v8 were faster:

```sh
v8

======================
 koa-router benchmark
======================
short static: 1,422,593 ops/sec
static with same radix: 1,385,510 ops/sec
dynamic route: 1,344,517 ops/sec
mixed static dynamic: 1,323,915 ops/sec
long static: 1,384,331 ops/sec
wildcard: 1,389,524 ops/sec
all together: 226,891 ops/sec


v7

======================
 koa-router benchmark
======================
short static: 1,498,510 ops/sec
static with same radix: 1,488,757 ops/sec
dynamic route: 1,465,601 ops/sec
mixed static dynamic: 1,419,721 ops/sec
long static: 1,457,910 ops/sec
wildcard: 1,422,492 ops/sec
all together: 237,266 ops/sec
```

*Something* happened between v8 and v9 that made things way better... And then v11 dialed it WAY back.

__Note:__ I had to change `koa-router`'s wildcard to a regex.

**Comment by jeremy-daley-kr** [comment]
Did some more digging from this comparison page: https://github.com/koajs/router/compare/v10.1.1...v11.0.0

Tried each commit, and seemed to pinpoint it to this commit:
https://github.com/koajs/router/commit/d0c6d8b430ca56081801637ae459caf68d5f022c

CC: @titanism 

Something in that commit made things VERY slow

**Comment by titanism** [comment]
This is an amazing find @jeremy-daley-kr (albeit I haven't reproduced yet nor tested benchmarks).

Perhaps in the commit it was the change in `for` loops, e.g. <https://github.com/koajs/router/commit/d0c6d8b430ca56081801637ae459caf68d5f022c#diff-9be572fb49ced6cd74dfd6787ba235abfa40f12b2c5abf47ac93a7a6d6cbc6c0L27-R28>?

**Reply by titanism** [reply]
@jeremy-daley-kr can you try editing in v11 tests the source code and revert those loops to have the same code as before?

**Reply by titanism** [reply]
@jeremy-daley-kr most likely related to this <https://stackoverflow.com/q/73148669>

**Reply by jeremy-daley-kr** [reply]
@titanism  I was actually wondering that too. `for of` always looks better, but I think comes with some overhead.

**Reply by jeremy-daley-kr** [reply]
@titanism Well, I've reverted every `for of` directly in node_modules, and don't see much of a difference. I wonder what else might be going on there.

**Reply by titanism** [reply]
I tried to fork that repo with benchmarks and run it but got this error for koa-router v10:

```sh
TypeError: Unexpected MODIFIER at 8, expected END
    at mustConsume (/Users/user/Projects/router-benchmark/node_modules/koa-router/node_modules/path-to-regexp/dist/index.js:114:15)
    at parse (/Users/user/Projects/router-benchmark/node_modules/koa-router/node_modules/path-to-regexp/dist/index.js:172:9)
    at stringToRegexp (/Users/user/Projects/router-benchmark/node_modules/koa-router/node_modules/path-to-regexp/dist/index.js:329:27)
    at pathToRegexp (/Users/user/Projects/router-benchmark/node_modules/koa-router/node_modules/path-to-regexp/dist/index.js:407:12)
    at new Layer (/Users/user/Projects/router-benchmark/node_modules/koa-router/lib/layer.js:43:17)
    at Router.register (/Users/user/Projects/router-benchmark/node_modules/koa-router/lib/router.js:570:17)
    at Router.<computed> [as get] (/Users/user/Projects/router-benchmark/node_modules/koa-router/lib/router.js:200:12)
    at /Users/user/Projects/router-benchmark/benchmarks/koa-router.js:30:12
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/Users/user/Projects/router-benchmark/benchmarks/koa-router.js:28:8)
```

```sh
node --version
v18.16.0
```

**Reply by jeremy-daley-kr** [reply]
Yeah... you have to change the last wildcard in `./benchmarks/koa-router.js`.

I went with this:

```js
{ method: 'GET', url: /^\/static\/.*$/ }
```

**Reply by jeremy-daley-kr** [reply]
Or you can just remove that wildcard test completely. The issue reveals itself before it gets to that benchmark anyway.

**Reply by jeremy-daley-kr** [reply]
@titanism I feel like I've tried everything obvious. I'm kinda dumbfounded on this one. FYI... The way I pinpointed that commit was by setting the `package.json` like so:

```json
"dependencies": {
  "koa-router": "git://github.com/koajs/router.git#d0c6d8b430ca56081801637ae459caf68d5f022c"
}
```

**Reply by titanism** [reply]
@jeremy-daley-kr if you figure out what's up please do let us know ❤️

**Reply by jeremy-daley-kr** [reply]
@titanism Do you think this should be submitted as a bug to the @koa/router repo? I'd like to get more community support, and make sure sight of this isn't lost.

**Reply by jeremy-daley-kr** [reply]
I believe I've at least ruled out `layer.js`. I copied/pasted the previous version of `layer.js` and `router.js`, and only saw the numbers shoot back up after reverting `router.js`.

**Reply by jeremy-daley-kr** [reply]
Found it!
It's this guy...

```js
const { debuglog } = require('util');
const debug = debuglog('koa-router');
```

Any particular reason you swapped that in for the previous:

```js
const debug = require('debug')('koa-router');
```

That single change appears to be the culprit.

**Reply by jeremy-daley-kr** [reply]
Pull request here:
https://github.com/koajs/router/pull/173

**Reply by titanism** [reply]
Wow, that's INSANE!  Native node debugger vs a third party package 🤦   Is there a GitHub issue about this in the Nodejs repo?

**Comment by titanism** [comment]
This is now resolved.  We have published a new version of v12 and v11 of `koa-router` and `@koa/router` (mirrored packages).

See <https://github.com/koajs/router/pull/173>.

Thank you sincerely @jeremy-daley-kr &ndash; we're going to get this into production at [Forward Email](https://forwardemail.net) today (GitHub at @forwardemail).
